src/nets/ctc_prefix_score.py

```diff
@@ -3,10 +3,142 @@
 # Copyright 2018 Mitsubishi Electric Research Labs (Takaaki Hori)
 #  Apache 2.0  (http://www.apache.org/licenses/LICENSE-2.0)
 
+import torch
+
 import numpy as np
+import cupy as np
 import six
 
 
+def logsumexp(inputs, dim=None, keepdim=False):
+    """Numerically stable logsumexp.
+
+    Args:
+    inputs: A Variable with any shape.
+    dim: An integer.
+    keepdim: A boolean.
+
+    Returns:
+    Equivalent of log(sum(exp(inputs), dim=dim, keepdim=keepdim)).
+    """
+    # For a 1-D array x (any array along a single dimension),
+    # log sum exp(x) = s + log sum exp(x - s)
+    # with s = max(x) being a common choice.
+    if dim is None:
+        inputs = inputs.view(-1)
+        dim = 0
+    s, _ = torch.max(inputs, dim=dim, keepdim=True)
+    outputs = s + (inputs - s).exp().sum(dim=dim, keepdim=True).log()
+    if not keepdim:
+        outputs = outputs.squeeze(dim)
+    return outputs
+
+
+class BatchCTCPrefixScoreTH(object):
```

Comment:
Same.
`BatchCTCPrefixScoreTH(object)` -> `CTCPrefixScoreTH(object)`, and replace the original one with the batch one.